MoltenVK/MoltenVK/Utility/MVKBitArray.h

```diff
@@ -104,29 +110,79 @@ class MVKBitArray {
 		return getIndexOfFirstSetBit(0, false);
 	}
 
+	/**
+	 * Enumerates the bits, executing a custom function on each bit that is enabled.
+	 *
+	 * The function to execute is passed a bitIndex parameter which indicates
+	 * the index of the bit for which the function is executing.
+	 *
+	 * The custom function should return true to continue processing further bits, or false
+	 * to stop processing further bits. This function returns false if any of the invocations
+	 * of the custom function halted further invocations, and returns true otherwise.
+	 *
+	 * If shouldClear is true, each enabled bit is cleared before the custom function executes.
+	 */
+	bool enumerateEnabledBits(bool shouldClear, std::function<bool(size_t bitIndex)> func) {
+		for (size_t bitIdx = getIndexOfFirstSetBit(shouldClear);
+			 bitIdx < _bitCount;
+			 getIndexOfFirstSetBit(++bitIdx, shouldClear)) {
+
+			if ( !func(bitIdx) ) { return false; }
+		}
+		return true;
+	}
+
 	/** Returns the number of bits in this array. */
 	inline size_t size() { return _bitCount; }
 
 	/** Returns whether this array is empty. */
 	inline bool empty() { return !_bitCount; }
 
-	/** Resize this array to the specified number of bits, and sets the initial value of all the bits. */
+	/**
+	 * Resize this array to the specified number of bits. The value of existing
+	 * bits that fit within the new size are retained, and any new bits that
+	 * are added to accommodate the new size are set to the given value.
+	 * Consumed memory is retained unless the size is set to zero.
+	 */
 	inline void resize(size_t size = 0, bool val = false) {
-		free(_pSections);
+		size_t oldBitCnt = _bitCount;
+		size_t oldSecCnt = getSectionCount();
 
 		_bitCount = size;
-		_pSections = _bitCount ? (uint64_t*)malloc(getSectionCount() * SectionByteCount) : nullptr;
-		if (val) {
-			setAllBits();
-		} else {
-			clearAllBits();
+		size_t newSecCnt = getSectionCount();
+
+		if (newSecCnt > oldSecCnt) {
+			uint64_t* pOldSecs = _pSections;
+			size_t oldByteCnt = oldSecCnt * SectionByteCount;
+			size_t newByteCnt = newSecCnt * SectionByteCount;
+
+			// Allocate new memory and fill it with the new initial value
+			_pSections = _bitCount ? (uint64_t*)malloc(newByteCnt) : nullptr;
+			if (_pSections) { memset(_pSections, val ? ~0 : 0, newByteCnt); }
+
+			// Copy the old contents to the new memory, and fill any bits in the old
+			// last section that were beyond the old bit count with the new initial value.
+			if (_pSections && pOldSecs) { memcpy(_pSections, pOldSecs, oldByteCnt); }
+			size_t oldEndBitCnt = oldSecCnt << SectionMaskSize;
+			for (size_t bitIdx = oldBitCnt; bitIdx < oldEndBitCnt; bitIdx++) { setBit(bitIdx, val); }

```

Comment:
The code here has been reworked, so this particular code here is obsolete. But, the reason that exists is because there are three steps to creating new memory: 

1. Allocate new memory and populate with the initial value.
2. Copy old memory to new. This is done in 8-byte sections, so in the case where the old `_bitCount` is not a multiple of 64, the old memory beyond the old `_bitCount` will have the old default value. Since this function allows the default value to be changed, it's possible that copying old to new will stomp over the new default value in these trailing bits with the old default value.
3. Ensure those maybe-stomped trailing bits have the new default value. It only iterates to the next 64-bit section boundary (the end of the old bits).

I know, I know. Why did I bother spending dev cycles on that rare contingency? 🤦🏻‍♂️ But it arose out of the use of `resize()` to do all memory creation and default population.